### PR TITLE
[SPARK-6309] [SQL] [MLlib] Implement MatrixUDT

### PR DESCRIPTION
**Pull request by MechCoder**
Utilities to serialize and deserialize Matrices in MLlib

### CODE REVIEW

**Comment by MechCoder** [comment]
cc @mengxr @rxin It would be great if you could have a look! Btw, should I wait for this to get merged, so that I can rebase the GaussianMixtureModel on top of this?

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 1]
```diff
@@ -20,12 +20,17 @@ package org.apache.spark.mllib.linalg
 import java.util.{Arrays, Random}
 
 import scala.collection.mutable.{ArrayBuilder => MArrayBuilder, HashSet => MHashSet, ArrayBuffer}
+import org.apache.spark.annotation.DeveloperApi
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: organize imports

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 2]
```diff
@@ -102,6 +107,83 @@ sealed trait Matrix extends Serializable {
...
+    // the dense matrix is built by numRows, numCols, values and isTransposed, all of which are
+    // set as not nullable. the sparse matrix needs colPtrs and rowIndices, which are set as
+    // null, while building the dense matrix.
+    StructType(Seq(
+      StructField("type", ByteType, nullable = false),
+      StructField("numRows", IntegerType, nullable = false),
+      StructField("numCols", IntegerType, nullable = false),
+      StructField("colPtrs", ArrayType(IntegerType, containsNull = false), nullable = true),
+      StructField("rowIndices", ArrayType(IntegerType, containsNull = false), nullable = true),
+      StructField("values", ArrayType(DoubleType, containsNull = false), nullable = false),
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: Make `values` nullable too. In the future, we might want to add binary matrix, where only indices are needed.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 3]
```diff
@@ -102,6 +107,83 @@ sealed trait Matrix extends Serializable {
...
+        row.setNullAt(4)
+        row.update(5, dm.values.toSeq)
+        row.setBoolean(6, dm.isTransposed)
+    }
+    row
+  }
+
+  override def deserialize(datum: Any): Matrix = {
+    datum match {
+      case m: Matrix => m
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: Put a TODO here. This should never happen, but we have a bug somewhere in the DataFrames implementation.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 4]
```diff
@@ -431,6 +515,13 @@ class SparseMatrix(
     }
   }
 
+  override def equals(o: Any) = o match {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: We need to override hashCode too. Let's move this out to another PR (with hashCode). We can merge that one first and then this UDT PR.

File: mllib/src/test/scala/org/apache/spark/mllib/linalg/MatricesSuite.scala [hunk 5]
```diff
@@ -424,4 +424,15 @@ class MatricesSuite extends FunSuite {
...
   }
+
+  test("MatrixUDT") {
+    val dm1 = new DenseMatrix(2, 2, Array(0.9, 1.2, 2.3, 9.8))
+    val dm2 = new DenseMatrix(0, 0, Array())
+    val sm1 = dm1.toSparse
+    val sm2 = dm2.toSparse
+    val mUDT = new MatrixUDT()
+    Seq(dm1, dm2, sm1, sm2).foreach {
+        mat => assert(mat == mUDT.deserialize(mUDT.serialize(mat)))
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/linalg/MatricesSuite.scala]: use `===` instead of `==`. The former contains more info in the error message.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala (hunk 3 above)
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: Oh, Could you give some code to reproduce this behavior? I just saw the VectorUDT implementation.

Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: I don't remember the exact code to re-produce this issue. This is before the DataFrames API, so things might have changed already. Please try removing this line in VectorUDT and running unit tests.

**Comment by mengxr** [comment]
@MechCoder Btw, please add new commits instead of squashing commits during the code review. It makes easier to see what you've changed in each update.

**Comment by MechCoder** [comment]
@mengxr I have removed equals and now the testing is done using conversion to Arrays.

Let us please not wait for the other PR and try to get this merged, Thanks!

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 6]
```diff
@@ -21,11 +21,17 @@ import java.util.{Arrays, Random}
 
 import scala.collection.mutable.{ArrayBuilder => MArrayBuilder, HashSet => MHashSet, ArrayBuffer}
 
+import org.apache.spark.annotation.DeveloperApi
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.catalyst.expressions.GenericMutableRow
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: organize imports (put this group under breeze)

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 7]
```diff
@@ -102,6 +108,86 @@ sealed trait Matrix extends Serializable {
...
+            new SparseMatrix(numRows, numCols, colPtrs, rowIndices, values, isTransposed)
+          case 1 =>
+            new DenseMatrix(numRows, numCols, values, isTransposed)
+        }
+    }
+  }
+
+  override def userClass: Class[Matrix] = classOf[Matrix]
+
+  override def equals(o: Any): Boolean = {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: Since we implement `equals`, let's add a `hashCode` that returns a predefined random integer. See

https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala#L186

(and you can pick a value;)

**Comment by mengxr** [comment]
LGTM except minor inline comments.

**Comment by MechCoder** [comment]
@mengxr fixed! I put the hash as my Year of Birth, so that my permanent mark is left on Spark :P

**Comment by karthikvadla16** [comment]
Hi,

Can anyone please share me a example for using MatrixUDT as datatype column in Spark SQL dataframe. I'm trying to load bunch of images into dataframe. Each image as matrix in dataframe column.

Really appreciate your help.

Thanks 
Karthik
